Sources/MongoSwift/MongoError.swift

```diff
@@ -1,6 +1,128 @@
+                                  result: _,
+                                  errorLabels: lhsErrorLabels),
+                  .bulkWriteError(writeErrors: rhsWriteErrors,
+                                  writeConcernError: rhsWCError,
+                                  result: _,
+                                  errorLabels: rhsErrorLabels)):
+            let cmp = { (l: BulkWriteError, r: BulkWriteError) in l.index < r.index }
+
+            return sortAndCompareOptionalArrays(lhs: lhsWriteErrors, rhs: rhsWriteErrors, cmp: cmp)
+                    && lhsWCError == rhsWCError
+                    && sortAndCompareOptionalArrays(lhs: lhsErrorLabels, rhs: rhsErrorLabels, cmp: { $0 < $1 })
+        default:
+            return false
+        }
+    }
+}
+
+/// The possible errors caused by improper use of the driver by the user.
+public enum UserError: MongoSwiftError, Equatable {
+    /// Thrown when the driver is incorrectly used.
+    case logicError(message: String)
+
+    /// Thrown when the user passes in invalid arguments to a driver method.
+    case invalidArgument(message: String)
+
+    public static func == (lhs: UserError, rhs: UserError) -> Bool {
+        switch (lhs, rhs) {
+        case (.logicError(message: _), .logicError(message: _)),
+             (.invalidArgument(message: _), .invalidArgument(message: _)):
+            return true
+        default:
+            return false
+        }
+    }
+}
+
+/// The possible errors that can occur unexpectedly during runtime.
+public enum RuntimeError: MongoSwiftError, Equatable {
+    /// Thrown when the driver encounters a internal error not caused by the user. This is usually indicative of a bug
+    /// or system related failure (e.g. during memory allocation).
+    case internalError(message: String)
+
+    /// Thrown when encountering a connection or socket related error.
+    /// May contain labels providing additional information on the nature of the error.
+    case connectionError(message: String, errorLabels: [String]?)
+
+    /// Thrown when encountering an authentication related error (e.g. invalid credentials).
+    case authenticationError(message: String)
+
+    public static func == (lhs: RuntimeError, rhs: RuntimeError) -> Bool {
+        switch (lhs, rhs) {
+        case (.internalError(message: _), .internalError(message: _)),
+             (.authenticationError(message: _), .authenticationError(message: _)):
+            return true
+        case let (.connectionError(message: _, errorLabels: lhsLabels),
+                  .connectionError(message: _, errorLabels: rhsLabels)):
+            return sortAndCompareOptionalArrays(lhs: lhsLabels, rhs: rhsLabels, cmp: { $0 < $1 })
+        default:
+            return false
+        }
+    }
+}
+
+/// Internal helper function used to get an appropriate error from a libmongoc error. This should NOT be used to get
+/// `.writeError`s or `.bulkWriteError`s.
```

Comment:
(maybe note how writeErrors and bulkWriteErrors should be gotten then)

Reply:
Done.